### PR TITLE
Initial QTI export of GI items

### PR DESCRIPTION
**Pull request by wtritch**
This PR introduces basic export GI items into QTI.

In the initial pass, I'm just exporting the content required to *render* the item.  
A future task will handle processing the machine-scoring.

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/CommonQtiMapperSupport.java [hunk 1]
```diff
@@ -50,4 +56,22 @@ public QtiOutcomeDeclaration createScoreOutcomeDeclaration(final Item item) {
...
+     * @return The associated QTI custom operator element
+     */
+    public CustomOperatorType createCustomOperator(final String type,
+                                                   final String function,
+                                                   final Map<String, String> definition) {
+        final CustomOperatorType customOperator = objectFactory.createCustomOperatorType();
+        customOperator.setClazz("qti.sbac.customOperators." + type + "." + function);
+        customOperator.setDefinition(DefinitionJoiner.join(definition));
+        return customOperator;
+    }
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/CommonQtiMapperSupport.java]: This is going to be needed by any item type that uses custom scoring operators, so I added it to the common support class.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EbsrQtiMapper.java [hunk 2]
```diff
@@ -49,7 +47,7 @@ public EbsrQtiMapper(final ObjectFactory objectFactory,
 
     @Override
     public int getOrder() {
-        return OrderMultipleChoiceQtiMapper;
+        return OrderEbsrQtiMapper;
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EbsrQtiMapper.java]: unrelated cleanup

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        if (!(context.getTimsItem() instanceof GiItem)) {
+            return;
+        }
+
+        final GiItem giItem = (GiItem) context.getTimsItem();
+        final QtiAssessmentItem target = context.getQtiItem();
+
+        target.getQtiResponseDeclarations().add(createResponseDeclaration());
+        target.getQtiOutcomeDeclarations().addAll(mapToOutcomeDeclarations(giItem));
+        target.setQtiResponseProcessing(createResponseProcessing(giItem));
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: Above is mapping all of the scoring elements.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+
+        final Div contentRoot = contentMapper.appendContentRoot(target);
+        final List content = contentRoot.getContent();
+        content.add(createStem(context, giItem));
+        content.add(createGiInteraction(context, giItem));
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: Above is mapping the prompt and GI interaction (GAX.)

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            .put(QName.valueOf("data-interaction-identifier"), "sbac:grid-item");
+
+        try {
+            final Document document = DocumentBuilderFactory.newInstance()
+                .newDocumentBuilder()
+                .newDocument();
+
+            final Element customOption = document.createElement("template");
+            customOption.appendChild(document.createCDATASection(giItem.getCore().getEn().getGiDefinition()));
+            customInteractionType.getAnies().add(customOption);
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: This is a little strange.  The CustomInteractionType created from the schema contains "any" child elements.  Unfortunately, the provided ObjectFactory doesn't have methods for creating generic `org.w3c.dom.Element` instances so we need to build them ourselves.

Comment (wtritch) [reply src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: @esotrope This is creating the:
```
<template>
  <![CDATA[<Question ... />]]>
</template>
```
element.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        mapToPreProcessResponseOutcomeVariable(scoring.ppResponseVariable())
+            .ifPresent(setValueType -> responseProcessing.getIncludesAndQtiResponseConditionsAndQtiResponseProcessingFragments()
+                .add(setValueType));
+
+        // TODO Add variable processing
+//        responseProcessing.getIncludesAndQtiResponseConditionsAndQtiResponseProcessingFragments()
+//            .addAll(scoring.getVariables().stream()
+//                .filter(variable -> !variable.equals(scoring.ppResponseVariable()))
+//                .map(this::mapToSetOutcomeValue)
+//                .collect(Collectors.toList()));
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: I started working on processing TIMS JSON machine-scoring definitions here, but will finish this off in a future PR.

It's a little awkward in that the TIMS representation is inverted from the QTI representation.  In TIMS a variable is declared with it's response-processing logic embedded.  So one TIMS Variable leads to two QTI elements:
1. The `<qti-outcome-definition>` that defines the baseType and cardinality of the variable
2. An associated `<qti-set-outcome-value>` tree within the `<qti-response-processing>` element that initializes the variable's value.  I'm really hoping we don't run into a case in the existing items where a variable can be set multiple ways/times over the course of response processing.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    private Optional<SetValueType> mapToPreProcessResponseOutcomeVariable(final Variable ppResponseVariable) {
+        if (ppResponseVariable == null) {
+            return Optional.empty();
+        }
+
+        final Map<String, String> definition = ImmutableMap.of(
+            "response", ResponseIdentifier
+        );
+        final CustomOperatorType ppResponseOperator = commonSupport.createCustomOperator("GRAPHIC", "PREPROCESSRESPONSE", definition);
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: This is implied/assumed within the TIMS json.  The "first" Variable is assumed to be "pre-processed" according to the associated item type.
In most existing items, this variable is named "PP_RESPONSE", in the TIMS json, it's hard-coded as "Responses".  We'll need to account for the renaming when we start ingesting QRX resources.  FYI @millerts

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+        final SetValueType setValueType = objectFactory.createSetValueType();
+        setValueType.setIdentifier(ppResponseVariable.getIdentifier());
+        setValueType.setQtiCustomOperator(ppResponseOperator);
+        return Optional.of(setValueType);
+    }
+
+    private SetValueType mapToSetOutcomeValue(final Variable variable) {
+        // TODO
+        return null;
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/GiQtiMapper.java]: This will need to perform the generic mapping if JSON operators to QTI operator elements.

File: src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EbsrQtiMapper.java (hunk 2 above)
Comment (agorina) [reply src/main/java/org/opentestsystem/ap/item/translation/mappers/qti/EbsrQtiMapper.java]: Thanks for fixing it!
